### PR TITLE
V2 Encryption Bug 4MB file downloads

### PR DESCRIPTION
**Pull request by ibrandes**
Addresses https://github.com/Azure/azure-sdk-for-java/issues/41709

### CODE REVIEW

File: sdk/storage/azure-storage-blob-cryptography/src/main/java/com/azure/storage/blob/specialized/cryptography/EncryptedBlobAsyncClient.java [hunk 1]
```diff
@@ -753,17 +757,44 @@ private <T> Mono<T> populateRequestConditionsAndContext(BlobRequestConditions re
...
+            EncryptionData encryptionData;
+            try (JsonReader jsonReader = JsonProviders.createReader(metadetaString)) {
+                encryptionData = EncryptionData.fromJson(jsonReader);
+            } catch (IOException e) {
+                throw new RuntimeException(e);
+            }
+            //if the blob is encrypted using V2, use the unencrypted length for range calculations
+            if ("2.0".equals(encryptionData.getEncryptionAgent().getProtocol())) {
+                long regionLength = getClientSideEncryptionOptions().getAuthenticatedRegionDataLengthInBytes();
+                long region = Math.floorDiv(totalLength, regionLength);
```
Comment (kyleknap) [inline sdk/storage/azure-storage-blob-cryptography/src/main/java/com/azure/storage/blob/specialized/cryptography/EncryptedBlobAsyncClient.java:783]: For this calculation, we should double check if we need to be doing floor or ceiling division. I'd imagine it would be ceiling as in the case where the total length fits in one encryption region, the total unencrypted length should be just the `total length - (nonce + tag)` and here it seems like it would just return the encrypted length?

Once we break this out to a separate class, we should be able to parameterize around these edge case in a unit test for the class to make sure we are confident it is behaving correctly.

File: sdk/storage/azure-storage-blob-cryptography/src/main/java/com/azure/storage/blob/specialized/cryptography/EncryptedBlobClient.java [hunk 2]
```diff
@@ -400,7 +407,8 @@ public Response<BlobProperties> downloadToFileWithResponse(String filePath, Blob
     @Override
     public Response<BlobProperties> downloadToFileWithResponse(BlobDownloadToFileOptions options, Duration timeout,
         Context context) {
-        context = context == null ? Context.NONE : context;
+        context = context == null ? new Context("rangeAdjustment", adjustRangeThroughContext())
```
Comment (kyleknap) [inline sdk/storage/azure-storage-blob-cryptography/src/main/java/com/azure/storage/blob/specialized/cryptography/EncryptedBlobClient.java:410]: For adding the adjusted length, we probably want to make the update in the `populateRequestConditionsAndContext` method similar to what we did in the async client.

File: sdk/storage/azure-storage-blob/src/main/java/com/azure/storage/blob/implementation/util/ChunkedDownloadUtils.java [hunk 3]
```diff
@@ -32,10 +34,11 @@ public class ChunkedDownloadUtils {
     Download the first chunk. Construct a Mono which will emit the total count for calculating the number of chunks,
     access conditions containing the etag to lock on, and the response from downloading the first chunk.
      */
+    @SuppressWarnings("unchecked")
```
Comment (kyleknap) [inline sdk/storage/azure-storage-blob/src/main/java/com/azure/storage/blob/implementation/util/ChunkedDownloadUtils.java:37]: What does this suppression do?

File: sdk/storage/azure-storage-blob/src/main/java/com/azure/storage/blob/specialized/BlobClientBase.java [hunk 4]
```diff
@@ -347,7 +347,8 @@ public BlobInputStream openInputStream(BlobInputStreamOptions options, Context c
             new com.azure.storage.common.ParallelTransferOptions().setBlockSizeLong((long) chunkSize);
         BiFunction<BlobRange, BlobRequestConditions, Mono<BlobDownloadAsyncResponse>> downloadFunc =
             (chunkRange, conditions) -> client.downloadStreamWithResponse(chunkRange, null, conditions, false, contextFinal);
-        return ChunkedDownloadUtils.downloadFirstChunk(range, parallelTransferOptions, requestConditions, downloadFunc, true)
```
Comment (kyleknap) [inline sdk/storage/azure-storage-blob/src/main/java/com/azure/storage/blob/specialized/BlobClientBase.java:350]: It might be worth double checking if `openInputStream` is impacted by the same issue as `downloadToFile`. I have not checked it, but would be good to double check since it is relying on the `ChunkedDownloadUtils` class and also see if we have any tests for this method in our tests.

File: sdk/storage/azure-storage-blob-cryptography/src/test/java/com/azure/storage/blob/specialized/cryptography/EncryptedBlockBlobApiTests.java [hunk 5]
```diff
@@ -1115,8 +1116,7 @@ private static Stream<Arguments> downloadFileSupplier() {
             Arguments.of(50 * Constants.MB, EncryptionVersion.V1), // large file requiring multiple requests
             Arguments.of(0, EncryptionVersion.V2), // empty file
             Arguments.of(20, EncryptionVersion.V2), // small file
-            //Arguments.of(16 * 1024 * 1024, EncryptionVersion.V2), // medium file in several chunks
-            //todo isbr: uncomment after 4mb boundary bug is fixed
+            Arguments.of(16 * 1024 * 1024, EncryptionVersion.V2), // medium file in several chunks
```
Comment (kyleknap) [inline sdk/storage/azure-storage-blob-cryptography/src/test/java/com/azure/storage/blob/specialized/cryptography/EncryptedBlockBlobApiTests.java:1119]: I think there may be a couple more tests that we'll need to uncomment like this one: https://github.com/Azure/azure-sdk-for-java/blame/27bef9ead66da9221f481063654c32829e189f99/sdk/storage/azure-storage-blob-cryptography/src/test/java/com/azure/storage/blob/specialized/cryptography/EncryptedBlockBlobApiTests.java#L1895. It may make sense to sync up with @ibrahimrabab to make sure we uncomment all of the test cases that we needed to comment out because of this issue.

File: sdk/storage/azure-storage-blob-changefeed/src/main/java/com/azure/storage/blob/changefeed/BlobChunkedDownloader.java [hunk 6]
```diff
@@ -52,7 +52,8 @@ public Flux<ByteBuffer> download() {
             -> client.downloadWithResponse(range, null, conditions, false);
 
         /* We don't etag lock since the Changefeed can append to the blob while we are reading it. */
-        return ChunkedDownloadUtils.downloadFirstChunk(range, options, requestConditions, downloadFunc, false)
+        return ChunkedDownloadUtils.downloadFirstChunk(range, options, requestConditions, downloadFunc, false,
+            null)
```
Comment (kyleknap) [inline sdk/storage/azure-storage-blob-changefeed/src/main/java/com/azure/storage/blob/changefeed/BlobChunkedDownloader.java:56]: In the `ChunkedDownloadUtils` class, I'm wondering if it makes sense to introduce this as an overloaded method? Mainly this will:
1. Allow us from avoiding needing to update the call pattern everywhere the chunked downloader class is used but no context is need to be provided
2. Help avoid conflict mismatches in clients (e.g. the customer has an older version of the blob client that does not support this method signature)

I'm not up to speed on how this is typically handled between packages so would be happy to hear thoughts on this idea.

**Review by kyleknap — COMMENTED** [review]
It is looking good. I like how we are able to minimize the invasiveness of the change with the context approach. I just had some general questions and suggestions related to refactoring some of this.

File: sdk/storage/azure-storage-blob-cryptography/src/main/java/com/azure/storage/blob/specialized/cryptography/EncryptedBlobAsyncClient.java [hunk 7]
```diff
@@ -753,17 +757,44 @@ private <T> Mono<T> populateRequestConditionsAndContext(BlobRequestConditions re
                     result = result.contextWrite(context -> context.put(ENCRYPTION_DATA_KEY,
                         EncryptionData.getAndValidateEncryptionData(encryptionDataKey, requiresEncryption)));
                 }
+                if (isDownloadToFile) {
+                    result = result.contextWrite(context -> context.put("rangeAdjustment", adjustRangeThroughContext()));
```
[the remark this answers is not in the document]
Comment (ibrandes) [reply sdk/storage/azure-storage-blob-cryptography/src/main/java/com/azure/storage/blob/specialized/cryptography/EncryptedBlobAsyncClient.java:761]: take this

File: sdk/storage/azure-storage-blob-cryptography/src/main/java/com/azure/storage/blob/specialized/cryptography/EncryptedBlobAsyncClient.java [hunk 8]
```diff
@@ -732,14 +736,14 @@ public Mono<BlobDownloadAsyncResponse> downloadStreamWithResponse(BlobRange rang
...
-                () -> super.downloadStreamWithResponse(range, options, requestConditions, getRangeContentMd5));
+                () -> super.downloadStreamWithResponse(range, options, requestConditions, getRangeContentMd5), false);
         } else {
             return super.downloadStreamWithResponse(range, options, requestConditions, getRangeContentMd5);
         }
     }
 
     private <T> Mono<T> populateRequestConditionsAndContext(BlobRequestConditions requestConditions,
-        Supplier<Mono<T>> downloadCall) {
+        Supplier<Mono<T>> downloadCall, Boolean isDownloadToFile) {
```
[the remark this answers is not in the document]
Comment (ibrandes) [reply sdk/storage/azure-storage-blob-cryptography/src/main/java/com/azure/storage/blob/specialized/cryptography/EncryptedBlobAsyncClient.java:746]: verify what apis use downloadfirstchunk and see if we can remove the boolean value

File: sdk/storage/azure-storage-blob/src/main/java/com/azure/storage/blob/implementation/util/ChunkedDownloadUtils.java (hunk 3 above)
Comment (ibrandes) [reply sdk/storage/azure-storage-blob/src/main/java/com/azure/storage/blob/implementation/util/ChunkedDownloadUtils.java:37]: There is no easy way to resolve the warning programmatically - and there are other instances in the SDK where Optional<T>.get is used with @SuppressWarnings("unchecked")

File: sdk/storage/azure-storage-blob-cryptography/src/main/java/com/azure/storage/blob/specialized/cryptography/EncryptedBlobAsyncClient.java (hunk 8 above)
[the remark this answers is not in the document]
Comment (ibrandes) [reply sdk/storage/azure-storage-blob-cryptography/src/main/java/com/azure/storage/blob/specialized/cryptography/EncryptedBlobAsyncClient.java:746]: all other download APIs in sync and async are safe to use with this change, as we only pass in the context to ChunkedDownloadUtils.downloadFirstChunk in downloadToFileImpl (which is only used by downloadtofileapis), and everywhere else that uses downloadFirstChunk has null passed into it.

File: sdk/storage/azure-storage-blob/src/main/java/com/azure/storage/blob/specialized/BlobClientBase.java (hunk 4 above)
Comment (ibrandes) [reply sdk/storage/azure-storage-blob/src/main/java/com/azure/storage/blob/specialized/BlobClientBase.java:350]: openInputStream is not impacted by the same issue, and works without the context adjustment passed in. I have added a test for it.

File: sdk/storage/azure-storage-blob-cryptography/src/main/java/com/azure/storage/blob/specialized/cryptography/EncryptedBlobAsyncClient.java [hunk 9]
```diff
@@ -753,17 +757,44 @@ private <T> Mono<T> populateRequestConditionsAndContext(BlobRequestConditions re
...
                         EncryptionData.getAndValidateEncryptionData(encryptionDataKey, requiresEncryption)));
                 }
+                if (isDownloadToFile) {
+                    result = result.contextWrite(context -> context.put("rangeAdjustment", adjustRangeThroughContext()));
+                }
                 return result;
             });
     }
 
+    private BiFunction<BlobDownloadAsyncResponse, Long, Long> adjustRangeThroughContext() {
```
[the remark this answers is not in the document]
Comment (ibrandes) [reply sdk/storage/azure-storage-blob-cryptography/src/main/java/com/azure/storage/blob/specialized/cryptography/EncryptedBlobAsyncClient.java]: make this change in a future PR

File: sdk/storage/azure-storage-blob-cryptography/src/test/java/com/azure/storage/blob/specialized/cryptography/EncryptedBlockBlobApiTests.java [hunk 10]
```diff
@@ -1966,4 +1969,138 @@ public Mono<HttpResponse> process(HttpPipelineCallContext context, HttpPipelineN
...
     }
+
+    private static Stream<Arguments> fourMBUploadsV2Supplier() {
+        return Stream.of(
+            Arguments.of(4194305),
+            Arguments.of(4194304),
+            Arguments.of(4194277),
+            Arguments.of(33554432)
+        );
+    }
```
Comment (kyleknap) [inline sdk/storage/azure-storage-blob-cryptography/src/test/java/com/azure/storage/blob/specialized/cryptography/EncryptedBlockBlobApiTests.java:1980]: Instead of creating new cases, could we add onto existing cases and leverage existing tests? For the sync client, it looks like we could add these cases to `downloadFile` test and the 16MB case to the `uploadAndDownloadToFileDifferentRegionLength` tests.

For the async tests, I'm not sure if I saw any async client tests so if there are not any we can use the cases that we have but instead it may make sense to use the parameterized cases from the synchronous versions

File: sdk/storage/azure-storage-blob/src/main/java/com/azure/storage/blob/implementation/util/ChunkedDownloadUtils.java [hunk 11]
```diff
@@ -100,6 +103,88 @@ && extractTotalBlobLength(blobStorageException.getResponse()
...
 
+    /*
+    Has a context value for additional download adjustments.
+
+    Download the first chunk. Construct a Mono which will emit the total count for calculating the number of chunks,
+    access conditions containing the etag to lock on, and the response from downloading the first chunk.
+     */
+    @SuppressWarnings("unchecked")
+    public static Mono<Tuple3<Long, BlobRequestConditions, BlobDownloadAsyncResponse>> downloadFirstChunk(
+        BlobRange range, ParallelTransferOptions parallelTransferOptions,
```
Comment (kyleknap) [inline sdk/storage/azure-storage-blob/src/main/java/com/azure/storage/blob/implementation/util/ChunkedDownloadUtils.java:114]: Ah so in creating this new public method instead of forking + adding the adjusted blob length logic, I was thinking we would also update the [original version](https://github.com/Azure/azure-sdk-for-java/pull/41813/files#diff-e54b10e2121ed9d4e7ec0eacf29f1feca0ae11a1294ac2f6b9d56bee8897b47eL35) of this method to call into this new version, passing in a context of `null`. It should be similar to how we handle sharing logic across overloaded methods in the `EncryptedBlobAsyncClient` class. The main reason I'm suggesting this is so that we can avoid maintaining two versions of the code and instead just have the original version of the method call this new version directly.

File: sdk/storage/azure-storage-blob-cryptography/src/main/java/com/azure/storage/blob/specialized/cryptography/EncryptedBlobLength.java [hunk 12]
```diff
@@ -0,0 +1,41 @@
...
+import static com.azure.storage.blob.specialized.cryptography.CryptographyConstants.NONCE_LENGTH;
+import static com.azure.storage.blob.specialized.cryptography.CryptographyConstants.TAG_LENGTH;
+
+/**
+ * This class provides helper methods for adjusting encrypted downloads.
+ * <p>
+ * RESERVED FOR INTERNAL USE.
+ */
+
+final class EncryptedBlobLength {
```
Comment (kyleknap) [inline sdk/storage/azure-storage-blob-cryptography/src/main/java/com/azure/storage/blob/specialized/cryptography/EncryptedBlobLength.java:22]: It would awesome if we could add unit tests for this new class to have a quick way to make sure our calculation of this adjusted blob range is correct, especially around some of the corner cases we ran into.

I think we should be able to do something similar to the parameterized tests in `EncryptedBlobRangeTests` and make a new `EncryptedBlobLengthTests` file with cases where we parameterize the encryption version, the encrypted length, and the region length and make sure the expected adjusted blob length is what we expect.

**Review by kyleknap — COMMENTED** [review]
It's looking good. Just had some smaller comments.

File: sdk/storage/azure-storage-blob-cryptography/src/main/java/com/azure/storage/blob/specialized/cryptography/EncryptedBlobLength.java [hunk 13]
```diff
@@ -0,0 +1,43 @@
...
+import static com.azure.storage.blob.specialized.cryptography.CryptographyConstants.NONCE_LENGTH;
+import static com.azure.storage.blob.specialized.cryptography.CryptographyConstants.TAG_LENGTH;
+
+/**
+ * This class provides helper methods for adjusting encrypted downloads.
+ */
+final class EncryptedBlobLength {
+    private static final ClientLogger LOGGER = new ClientLogger(EncryptedBlobLength.class);
+
+    static Function<Long, Long> computeAdjustedBlobLength(EncryptionData encryptionData) {
```
Comment (kyleknap) [inline sdk/storage/azure-storage-blob-cryptography/src/main/java/com/azure/storage/blob/specialized/cryptography/EncryptedBlobLength.java:22]: I think I missed this in the last review, but is there a particular reason we need to return a function here? Could we instead just return the long directly? I think it would be easier to follow the code if this was a long instead. It was sort of difficult following the `apply()` call that happens in the downstream ChunkedDownloadUtil logic.

**Review by kyleknap — COMMENTED** [review]
Looks good. I just had a couple more small questions/comments.

**Review by kyleknap — COMMENTED** [review]
Looks good. I just had a few more small comments. We should also probably add to our changelog that we fixed the same issue in the `openInputStream` API.

File: sdk/storage/azure-storage-blob/src/main/java/com/azure/storage/blob/specialized/BlobInputStream.java [hunk 14]
```diff
@@ -97,4 +99,14 @@ public BlobProperties getProperties() {
         return this.properties;
     }
 
+    /**
+     * Allows for encrypted blobs to use BlobInputStream correctly by using the non-encrypted blob length
+     */
+    private static Long adjustBlobLength(Long initialLength, Context context) {
+        if (context != null && context.getData(Constants.ADJUSTED_BLOB_LENGTH_KEY).isPresent()) {
+            return (Long) context.getData(Constants.ADJUSTED_BLOB_LENGTH_KEY).get();
```
Comment (kyleknap) [inline sdk/storage/azure-storage-blob/src/main/java/com/azure/storage/blob/specialized/BlobInputStream.java:107]: Should we be using `long` instead of `Long` here? Looking over the `StorageInputStream` constructor and `getBlobSize()` method, they both use/return `long`. So it probably makes sense to make it consistent.

**Review by kyleknap — APPROVED** [review]
Looks good. Good catch on finding those other decryption issues :ship:
